logger/mask.go

```diff
@@ -0,0 +1,196 @@
+package logger
+
+import (
+	"strings"
+
+	"github.com/antchfx/xmlquery"
+	"github.com/valyala/fastjson"
+	"github.com/wego/pkg/collection"
+	"github.com/wego/pkg/errors"
+)
+
+// MaskData the data as well as the information on how to mask the data
+type MaskData struct {
+	FistCharsToShow int
+	LastCharsToShow int
+	CharsToIgnore   []rune
+	XMLTag          string
+	JSONKeys        []string
+}
+
+// MaskXML masks parts of the inner text of tags from the input XML with replacement
+func MaskXML(xml, maskChar string, toMasks []MaskData) string {
+	maskChar = maskCharOrDefault(maskChar)
+
+	doc, err := xmlquery.Parse(strings.NewReader(xml))
+	if err != nil {
+		return errors.New("invalid XML input", err).Error()
+	}
+	out := findTagAndMaskMulti(doc, maskChar, toMasks)
+
+	return out
+}
+
+func findTagAndMaskMulti(doc *xmlquery.Node, maskChar string, toMasks []MaskData) string {
+	for _, toMask := range toMasks {
+		findTagAndMask(doc, maskChar, toMask)
+	}
+	return doc.OutputXML(true)
+}
+
+func findTagAndMask(doc *xmlquery.Node, maskChar string, toMask MaskData) {
+	nodes := xmlquery.Find(doc, "//"+toMask.XMLTag)
+	for _, node := range nodes {
+		tagValue := node.InnerText()
+		if strings.TrimSpace(tagValue) != "" {
+			var nodeToUpdate *xmlquery.Node
+			// get specific node that contains the value to update
+			for child := node.FirstChild; child != nil; child = child.NextSibling {
+				nodeToUpdate = child
+			}
+			maskedVal := getMaskedValue(maskChar, tagValue, toMask)
+			nodeToUpdate.Data = maskedVal
```

Comment:
I think we can also apply this to RedactXML to redact the xml values while iterating, instead of doing `replaceAll` on the whole xml string per replacement.